### ISSUE TITLE
running multiple JVM instances causing Loader to fail?

### ISSUE THREAD

**Issue by wongjoel** [bytedeco/javacpp#197]
Java CPP version: 1.3.2
JDK Version: 1.8.0_141
OS: Windows 7 x64

After recently updating a Jenkins instance that builds projects using JavaCV, I noticed that certain unit tests for the builds were failing intermittently. Looking at the console log, these failures seem to come down to a FileNotFoundException when the loader tries to load a dll (e.g. jniopencv_imgproc.dll). Seeing the message `(The process cannot access the file because it is being used by another process)` leads me to guess that this only happens when multiple builds (and tests) are happening at the same time, though I'm not sure why this was not happening before the update to Jenkins.

Just wondering if this is expected behaviour? In particular, do I need to do something like explicitly change the name the cache dir to something unique to avoid this kind of intermittent failure?

Thanks,

-------------
(Attached a partial typical stack trace of the problem)
```
Caused by: java.lang.UnsatisfiedLinkError: java.io.FileNotFoundException: C:\Users\username\.javacpp\cache\opencv-3.2.0-1.3-windows-x86_64.jar\org\bytedeco\javacpp\windows-x86_64\jniopencv_imgproc.dll (The process cannot access the file because it is being used by another process)
	at org.bytedeco.javacpp.Loader.loadLibrary(Loader.java:983)
	at org.bytedeco.javacpp.Loader.load(Loader.java:764)
	at org.bytedeco.javacpp.Loader.load(Loader.java:671)
	at org.bytedeco.javacpp.opencv_imgproc.<clinit>(opencv_imgproc.java:12)
	at java.lang.Class.forName0(Native Method)
	at java.lang.Class.forName(Class.java:348)
	at org.bytedeco.javacpp.Loader.load(Loader.java:726)
	at org.bytedeco.javacpp.Loader.load(Loader.java:671)
	at org.bytedeco.javacpp.opencv_imgcodecs.<clinit>(opencv_imgcodecs.java:13)
```

**Comment by saudet** [comment]
Duplicate of #182 (pull #188)

**Comment by wongjoel** [comment]
I see - I hadn't realised that the fix isn't yet available in a released version of JavaCPP. Thanks.

**Comment by saudet** [comment]
It is now in JavaCPP 1.3.3.

**Comment by wongjoel** [comment]
Updated to JavaCV 1.3.3, and apparently this bug persists. Same stack trace, but now the line numbers are shifted down slightly due to the extra code added to fix the other bug.

Somewhat annoying, but the only time I've ever seen this issue is when Jenkins is running tests on multiple builds at once, so perhaps not a big deal?

From the stack trace, I can see where the exception is first caught, but I can't work out which line in the try block is the source of the exception.

```
Caused by: java.lang.UnsatisfiedLinkError: java.io.FileNotFoundException: C:\Users\DEMC-Backup\.javacpp\cache\opencv-3.2.0-1.3-windows-x86_64.jar\org\bytedeco\javacpp\windows-x86_64\jniopencv_imgproc.dll (The process cannot access the file because it is being used by another process)
	at org.bytedeco.javacpp.Loader.loadLibrary(Loader.java:996)
	at org.bytedeco.javacpp.Loader.load(Loader.java:777)
	at org.bytedeco.javacpp.Loader.load(Loader.java:684)
	at org.bytedeco.javacpp.opencv_imgproc.<clinit>(opencv_imgproc.java:12)
	at java.lang.Class.forName0(Native Method)
	at java.lang.Class.forName(Class.java:348)
	at org.bytedeco.javacpp.Loader.load(Loader.java:739)
	at org.bytedeco.javacpp.Loader.load(Loader.java:684)
	at org.bytedeco.javacpp.opencv_imgcodecs.<clinit>(opencv_imgcodecs.java:13)
```

**Comment by saudet** [comment]
Can you check that a file named ".lock" appears in the root of
"~/.javacpp/cache"?

**Comment by wongjoel** [comment]
yes, there is a file named ".lock" in ~/.javacpp/cache, and the creation and modification times make sense.

**Comment by saudet** [comment]
And how does the test code below behave?
https://github.com/bytedeco/javacpp/issues/182#issuecomment-302663916

**Comment by wongjoel** [comment]
I can't get that code to throw any errors on either 1.3.3 or 1.3.2. I suppose that means I need to think harder about how to run two JVMs at exactly the same time, or submit my own test case.

**Comment by wongjoel** [comment]
Okay, I have now been able to replicate something similar to the #182 behavior for both 1.3.2 and 1.3.3:

In 1.3.2, the output (same for both JVMs) is:
```
Loading class org.bytedeco.javacpp.opencv_core
Loading class org.bytedeco.javacpp.opencv_core
Extracting jar:file:/C:/Users/username/Documents/git/skycam-utils/target/skycam-utils-0.0.3-bin/skycam-utils-0.0.3/lib/opencv-3.2.0-1.3-windows-x86_64.jar!/org/bytedeco/javacpp/windows-x86_64/msvcr120.dll
Loading C:\Users\username\.javacpp\cache\opencv-3.2.0-1.3-windows-x86_64.jar\org\bytedeco\javacpp\windows-x86_64\msvcr120.dll
Extracting jar:file:/C:/Users/username/Documents/git/skycam-utils/target/skycam-utils-0.0.3-bin/skycam-utils-0.0.3/lib/opencv-3.2.0-1.3-windows-x86_64.jar!/org/bytedeco/javacpp/windows-x86_64/msvcp120.dll
Loading C:\Users\username\.javacpp\cache\opencv-3.2.0-1.3-windows-x86_64.jar\org\bytedeco\javacpp\windows-x86_64\msvcp120.dll
Extracting jar:file:/C:/Users/username/Documents/git/skycam-utils/target/skycam-utils-0.0.3-bin/skycam-utils-0.0.3/lib/opencv-3.2.0-1.3-windows-x86_64.jar!/org/bytedeco/javacpp/windows-x86_64/opencv_core320.dll
Loading C:\Users\username\.javacpp\cache\opencv-3.2.0-1.3-windows-x86_64.jar\org\bytedeco\javacpp\windows-x86_64\opencv_core320.dll
Failed to load C:\Users\username\.javacpp\cache\opencv-3.2.0-1.3-windows-x86_64.jar\org\bytedeco\javacpp\windows-x86_64\opencv_core320.dll: java.lang.UnsatisfiedLinkError: C:\Users\username\.javacpp\cache\opencv-3.2.0-1.3-windows-x86_64.jar\org\bytedeco\javacpp\windows-x86_64\opencv_core320.dll: The process cannot access the file because it is being used by another process
Loading C:\Users\username\.javacpp\cache\opencv-3.2.0-1.3-windows-x86_64.jar\org\bytedeco\javacpp\windows-x86_64\opencv_core320.dll
Extracting jar:file:/C:/Users/username/Documents/git/skycam-utils/target/skycam-utils-0.0.3-bin/skycam-utils-0.0.3/lib/opencv-3.2.0-1.3-windows-x86_64.jar!/org/bytedeco/javacpp/windows-x86_64/opencv_imgproc320.dll
Loading C:\Users\username\.javacpp\cache\opencv-3.2.0-1.3-windows-x86_64.jar\org\bytedeco\javacpp\windows-x86_64\opencv_imgproc320.dll
Extracting jar:file:/C:/Users/username/Documents/git/skycam-utils/target/skycam-utils-0.0.3-bin/skycam-utils-0.0.3/lib/opencv-3.2.0-1.3-windows-x86_64.jar!/org/bytedeco/javacpp/windows-x86_64/jniopencv_core.dll
Loading C:\Users\username\.javacpp\cache\opencv-3.2.0-1.3-windows-x86_64.jar\org\bytedeco\javacpp\windows-x86_64\jniopencv_core.dll
Failed to load C:\Users\username\.javacpp\cache\opencv-3.2.0-1.3-windows-x86_64.jar\org\bytedeco\javacpp\windows-x86_64\jniopencv_core.dll: java.lang.UnsatisfiedLinkError: C:\Users\username\.javacpp\cache\opencv-3.2.0-1.3-windows-x86_64.jar\org\bytedeco\javacpp\windows-x86_64\jniopencv_core.dll: The process cannot access the file because it is being used by another process
Loading C:\Users\username\.javacpp\cache\opencv-3.2.0-1.3-windows-x86_64.jar\org\bytedeco\javacpp\windows-x86_64\jniopencv_core.dll
Loading class org.bytedeco.javacpp.opencv_core
Registering org.bytedeco.javacpp.opencv_core$CvSlice[address=0x437880,position=0,limit=1,capacity=1,deallocator=org.bytedeco.javacpp.Pointer$NativeDeallocator[ownerAddress=0x437880,deallocatorAddress=0x7febc9b55a0]]
Registering org.bytedeco.javacpp.opencv_core$CvSlice[address=0x437890,position=0,limit=1,capacity=1,deallocator=org.bytedeco.javacpp.Pointer$NativeDeallocator[ownerAddress=0x437890,deallocatorAddress=0x7febc9b55a0]]
Loading class org.bytedeco.javacpp.opencv_core
Registering org.bytedeco.javacpp.opencv_core$Mat[width=100,height=200,depth=8,channels=3]
100 200
```
In 1.3.3, the output (same for both JVMs) is:
```
Loading class org.bytedeco.javacpp.opencv_core
Loading class org.bytedeco.javacpp.opencv_core
Locking C:\Users\username\.javacpp\cache
Extracting jar:file:/C:/Users/username/Documents/git/skycam-utils/target/skycam-utils-0.0.3-bin/skycam-utils-0.0.3/lib/opencv-3.2.0-1.3-windows-x86_64.jar!/org/bytedeco/javacpp/windows-x86_64/msvcr120.dll
Loading C:\Users\username\.javacpp\cache\opencv-3.2.0-1.3-windows-x86_64.jar\org\bytedeco\javacpp\windows-x86_64\msvcr120.dll
Locking C:\Users\username\.javacpp\cache
Loading C:\Users\username\.javacpp\cache\opencv-3.2.0-1.3-windows-x86_64.jar\org\bytedeco\javacpp\windows-x86_64\msvcp120.dll
Locking C:\Users\username\.javacpp\cache
Extracting jar:file:/C:/Users/username/Documents/git/skycam-utils/target/skycam-utils-0.0.3-bin/skycam-utils-0.0.3/lib/opencv-3.2.0-1.3-windows-x86_64.jar!/org/bytedeco/javacpp/windows-x86_64/opencv_core320.dll
Loading C:\Users\username\.javacpp\cache\opencv-3.2.0-1.3-windows-x86_64.jar\org\bytedeco\javacpp\windows-x86_64\opencv_core320.dll
Locking C:\Users\username\.javacpp\cache
Loading C:\Users\username\.javacpp\cache\opencv-3.2.0-1.3-windows-x86_64.jar\org\bytedeco\javacpp\windows-x86_64\opencv_imgproc320.dll
Locking C:\Users\username\.javacpp\cache
Extracting jar:file:/C:/Users/username/Documents/git/skycam-utils/target/skycam-utils-0.0.3-bin/skycam-utils-0.0.3/lib/opencv-3.2.0-1.3-windows-x86_64.jar!/org/bytedeco/javacpp/windows-x86_64/jniopencv_core.dll
Loading C:\Users\username\.javacpp\cache\opencv-3.2.0-1.3-windows-x86_64.jar\org\bytedeco\javacpp\windows-x86_64\jniopencv_core.dll
Loading class org.bytedeco.javacpp.opencv_core
Registering org.bytedeco.javacpp.opencv_core$CvSlice[address=0x2da180,position=0,limit=1,capacity=1,deallocator=org.bytedeco.javacpp.Pointer$NativeDeallocator[ownerAddress=0x2da180,deallocatorAddress=0x7febca955a0]]
Registering org.bytedeco.javacpp.opencv_core$CvSlice[address=0x2da190,position=0,limit=1,capacity=1,deallocator=org.bytedeco.javacpp.Pointer$NativeDeallocator[ownerAddress=0x2da190,deallocatorAddress=0x7febca955a0]]
Loading class org.bytedeco.javacpp.opencv_core
Registering org.bytedeco.javacpp.opencv_core$Mat[width=100,height=200,depth=8,channels=3]
100 200
```
So it would appear that the fix is doing something. It also appears that the behaviour of Windows is different to the behaviour of Linux for this case.

**Comment by wongjoel** [comment]
One difference I've noticed between the issue I'm having and the test code from #182 is that this issue happens every time, whereas the test code from #182 only seems to fail on the first run after deleting the cache directory.

For what it's worth, the error I'm seeing has so far only involved `jniopencv_imgproc.dll`. It seems to try to extract it on every run. Should we expect the loader to attempt to extract it every run?

I've attached the output from my issue with the JavaCV debug output turned on
```
...
Loading class org.bytedeco.javacpp.opencv_core
Loading class org.bytedeco.javacpp.opencv_imgproc
Locking C:\Users\username\.javacpp\cache
Extracting jar:file:/C:/Users/username/.m2/repository/org/bytedeco/javacpp-presets/opencv/3.2.0-1.3/opencv-3.2.0-1.3-windows-x86_64.jar!/org/bytedeco/javacpp/windows-x86_64/jniopencv_imgproc.dll
Failed to extract for jniopencv_imgproc: java.lang.UnsatisfiedLinkError: java.io.FileNotFoundException: C:\Users\username\.javacpp\cache\opencv-3.2.0-1.3-windows-x86_64.jar\org\bytedeco\javacpp\windows-x86_64\jniopencv_imgproc.dll (The process cannot access the file because it is being used by another process)
Loading class org.bytedeco.javacpp.opencv_core
Loading class org.bytedeco.javacpp.opencv_imgproc
Registering (10, 10)
Registering org.bytedeco.javacpp.opencv_core$IplImage[width=10,height=10,depth=8,channels=1]
Registering (10, 10)
Registering org.bytedeco.javacpp.opencv_core$IplImage[width=10,height=10,depth=8,channels=1]
Registering (255.0, 0.0, 0.0, 0.0)
Registering (0.0, 0.0, 0.0, 0.0)
{0; 1} rotated around {1; 1} by 90.0�= {1; 2}Tests run: 6, Failures: 0, Errors: 3, Skipped: 0, Time elapsed: 1.234 sec <<< FAILURE!
lensGridProjectionTest(JavaCVTest)  Time elapsed: 0.963 sec  <<< ERROR!
java.lang.UnsatisfiedLinkError: java.io.FileNotFoundException: C:\Users\username\.javacpp\cache\opencv-3.2.0-1.3-windows-x86_64.jar\org\bytedeco\javacpp\windows-x86_64\jniopencv_imgproc.dll (The process cannot access the file because it is being used by another process)
	at org.bytedeco.javacpp.Loader.loadLibrary(Loader.java:996)
	at org.bytedeco.javacpp.Loader.load(Loader.java:777)
	at org.bytedeco.javacpp.Loader.load(Loader.java:684)
	at org.bytedeco.javacpp.opencv_imgproc.<clinit>(opencv_imgproc.java:12)
...
Caused by: java.io.FileNotFoundException: C:\Users\username\.javacpp\cache\opencv-3.2.0-1.3-windows-x86_64.jar\org\bytedeco\javacpp\windows-x86_64\jniopencv_imgproc.dll (The process cannot access the file because it is being used by another process)
	at java.io.FileOutputStream.open0(Native Method)
	at java.io.FileOutputStream.open(FileOutputStream.java:270)
	at java.io.FileOutputStream.<init>(FileOutputStream.java:213)
	at java.io.FileOutputStream.<init>(FileOutputStream.java:162)
	at org.bytedeco.javacpp.Loader.extractResource(Loader.java:574)
	at org.bytedeco.javacpp.Loader.cacheResource(Loader.java:438)
	at org.bytedeco.javacpp.Loader.loadLibrary(Loader.java:944)
	... 35 more
```

**Comment by saudet** [comment]
No, it checks the size and the date, and if they are the same, the file
isn't overwritten. Maybe it has something to do with permissions on your
installation of Windows? Like not allowing changing the dates of files?

**Comment by wongjoel** [comment]
Hmm, pinning down the symptoms is turning out to be harder than I initially thought.

When running with only one JVM, the files will be extracted on the first run, and loaded from cache on subsequent runs (as expected).

when running with two JVMs, I've observed the all of the following behaviours:
- files extracted first run, and then not being extracted in subsequent runs (expected behaviour)
- files (incorrectly) extracted after first run, but without failing (lock acquired, and extracts fine)
- files (incorrectly) extracted after first run, and failing as described previously (lock acquired, failed to extract)

It would seem that the presence of the second JVM is somehow capable of causing the check to see if the file has already been extracted to falsely indicate that the file is not extracted when it actually has been.

The behaviour is the same both when running in a standard command prompt and in an administrator command prompt. The "date modified" field for the *.dll files is set to be in the past, months before the creation date of today, so that part would seem to be working too.

**Comment by saudet** [comment]
So it's a problem with the Windows API?

**Comment by wongjoel** [comment]
I'm not sufficiently familiar with how it works to know how to confirm, but that does seem to be a fair explanation.

**Comment by saudet** [comment]
In any case, we need to figure out how to make it work, so please do continue your experiments! Thanks a lot

**Comment by saudet** [comment]
I've improved various things with the loader in commit https://github.com/bytedeco/javacpp/commit/8b05de89b91ded6287bbddd40caed6ecdce11cdd and it seems to have fixed whatever the problem was. I'm not able to reproduce any of those issues on Windows anymore. Could you give it a try? Unless you're still getting exceptions, I'll be considering this to be fixed. Thanks for the feedback!

**Comment by wongjoel** [comment]
Thanks - I shall attempt to try again with the new build

**Comment by saudet** [comment]
The newly released JavaCPP 1.4 contains these fixes, and given the lack of feedback I'm assuming all is working well now, but please let me know if this isn't the case! Thanks

**Comment by saudet** [comment]
Nope, this is still happening with Jenkins for example here: https://jenkins.ci.skymind.io/blue/organizations/jenkins/deeplearing4j%2Fdeeplearning4j/detail/PR-6204/12/pipeline/70

@wongjoel If you can provide a code snippet to reproduce with high probability, please do! Thanks

**Comment by saudet** [comment]
Latest fixes are now included in JavaCPP 1.5.1, which should prevent such conflicts from occurring again. Thanks again for your help!
